pkg/operatorstatus/monitor.go

```diff
@@ -0,0 +1,215 @@
+
+	// minimumSyncs represents the minimum number of syncs Marketplace must see
+	// prior to reporting that it is available to the ClusterOperator.
+	minimumSyncs uint = 2
+
+	// degradedRatio is the ratio of failed transition to syncs that must be reached
+	// prior to reporting that the marketplace operator is  degraded.
+	degradedRatio = 0.7
+)
+
+var (
+	instance *statusMonitor
+	once     sync.Once
+)
+
+type statusMonitor struct {
+	coWriter     *writer
+	namespace    string
+	eventTracker eventTracker
+	eventCh      chan error
+}
+
+// initInstance initializes the statusMonitor singleton named instance.
+func initInstance() error {
+	// Get the watch namespace.
+	namespace, err := k8sutil.GetWatchNamespace()
+	if err != nil {
+		return fmt.Errorf("failed to get watch namespace: %v", err)
+	}
+
+	// Create a new writer.
+	coWriter, err := newWriter()
+	if err != nil {
+		return fmt.Errorf("Error creating ClusterOpreator writer: %v", err)
+	}
+
+	// Initialize the instance singleton once.
+	once.Do(func() {
+		// init the instance singleton.
+		instance = &statusMonitor{
+			coWriter:     coWriter,
+			namespace:    namespace,
+			eventTracker: eventTracker{0, 0, 1000, 10, sync.Mutex{}},
+			eventCh:      make(chan error, 32),
+		}
+	})
+
+	return nil
+}
+
+// StartReporting initializes the instance singleton and starts reporting
+// ClusterOperator status.
+func StartReporting() error {
+	err := initInstance()
+	if err != nil {
+		return err
+	}
+
+	err = instance.reportProgressing()
+	if err != nil {
+		log.Errorf("[status] Error updating status: %v", err)
+	}
+
+	go instance.eventChannelReceiver()
+	go instance.monitorEvents()
+
+	return nil
+}
+
+// eventChannelReceiver will listen on the event channel and update the
+// number of syncs and failed phase transition events appropriately.
+func (s *statusMonitor) eventChannelReceiver() {
+	log.Info("[status] Starting event receiver")
+	for {
+		select {
+		case err := <-s.eventCh:
+			if err == nil {
+				s.eventTracker.incrementSyncs()
+			} else {
+				s.eventTracker.incrementFailedTransitions()
+			}
+		}
+	}
+}
+
+// SendEventMessage is used to send events to the eventCh. If the channel is
+// busy, the event will be dropped to prevent the controller from stalling.
+func SendEventMessage(err error) {
+	// If the coAPI is not available do not attempt to send messages to the
+	// sync channel
+	if instance == nil {
+		return
+	}
+
+	// A missing sync status is better than stalling the controller
+	select {
+	case instance.eventCh <- err:
+		break
+	default:
+		log.Warning("[status] Event channel is busy, not reporting event")
+	}
+}
+
+// monitorEvents will update the ClusterOpreator to reflect that Marketplace
+// is available or degraded based on the events reported.
+func (s *statusMonitor) monitorEvents() {
+	// reportInterval represents the frequency at which marketplace
+	// will attempt to update the ClusterOperator.
+	reportInterval := 1 * time.Minute
```

Comment:
This should be a constant

Reply:
If this is a constant we won't be able to change the interval at which we update the `ClusterOperator`.

Reply:
Rather than update the interval, I could replace the code where the value is updated with a `sleep`.

Reply:
> If this is a constant we won't be able to change the interval at which we update the `ClusterOperator`.

`1 * time.Minute` is a constant

>Rather than update the interval, I could replace the code where the value is updated with a sleep

You should use a sleep to avoid busy polling but either way the interval should increase.